File: minestom4fun-api/java/src/main/java/org/kanelucky/api/java/world/spawner/SpawnRule.java
package org.kanelucky.api.java.world.spawner;

import net.minestom.server.coordinate.Point;
import net.minestom.server.entity.Entity;
import net.minestom.server.instance.Instance;
import net.minestom.server.registry.RegistryKey;
import net.minestom.server.world.biome.Biome;

import java.util.Set;

/**
 * @author Kanelucky
 */
public interface SpawnRule {
    int getWeight();
    int getMinGroupSize();
    int getMaxGroupSize();
    Set<RegistryKey<Biome>> getValidBiomes();

    boolean canSpawn(Instance instance, Point pos);
    Entity createEntity();

    default boolean isValidSpawnPos(Instance instance, Point pos) {
        var chunk = instance.getChunkAt(pos);
        if (chunk == null || !chunk.isLoaded()) return false;

        var ground = instance.getBlock(pos.blockX(), pos.blockY() - 1, pos.blockZ());
        if (!ground.isSolid()) return false;

        var at = instance.getBlock(pos.blockX(), pos.blockY(), pos.blockZ());
        var above = instance.getBlock(pos.blockX(), pos.blockY() + 1, pos.blockZ());
        if (at.isSolid() || above.isSolid()) return false;

        int skyLight = instance.getSkyLight(pos.blockX(), pos.blockY(), pos.blockZ());
        if (skyLight < 9) return false;

        var biome = instance.getBiome(pos.blockX(), pos.blockY(), pos.blockZ());
        return getValidBiomes().contains(biome);
    }
}

File: minestom4fun-server/src/main/kotlin/org/kanelucky/server/world/spawners/SpawnRule.kt
package org.kanelucky.server.world.spawners

import net.minestom.server.coordinate.Point
import net.minestom.server.instance.Instance
import net.minestom.server.instance.block.Block

/**
 * Defines natural spawn conditions for an entity type
 *
 * @author Kanelucky
 */
interface SpawnRule {
    val weight: Int
    val minGroupSize: Int
    val maxGroupSize: Int
    fun canSpawn(instance: Instance, pos: Point): Boolean
    fun createEntity(): net.minestom.server.entity.Entity

    fun isGrass(instance: Instance, pos: Point): Boolean {
        val chunk = instance.getChunkAt(pos) ?: return false
        if (!chunk.isLoaded) return false
        val ground = instance.getBlock(pos.blockX(), pos.blockY() - 1, pos.blockZ())
        val above = instance.getBlock(pos.blockX(), pos.blockY(), pos.blockZ())
        return ground == Block.GRASS_BLOCK && !above.isSolid
    }
}